File: client/packages/vance-face/src/document/mapCodec.ts
// Codec for `kind: map` documents — parses an on-disk body into a
// typed MapDocument and serializes it back. JSON and YAML only;
// markdown is intentionally not supported.
//
// Mirrors the Java MapCodec under
// `vance-shared/src/main/java/de/mhus/vance/shared/document/kind/MapCodec.java`.
// On-disk format is documented in `specification/doc-kind-map.md`.
//
// Position is carried flat on markers: `place` / `lat` / `lon` at
// the marker level, not nested under a `location:` sub-object. Area
// `points` and route `waypoints` are arrays of the same flat shape.

import {
  dumpYamlBody,
  parseYamlBody,
  unwrapJsonMeta,
  wrapJsonMeta,
} from '@vance/shared';

export interface MapLocation {
  place?: string;
  lat?: number;
  lon?: number;
}

export interface MapMarker {
  name: string;
  title?: string;
  location: MapLocation;
  color?: string;
  description?: string;
  /** Unknown per-marker fields, preserved across round-trip. */
  extra: Record<string, unknown>;
}

export interface MapArea {
  name: string;
  title?: string;
  points: MapLocation[];
  color?: string;
  fillOpacity?: number;
  extra: Record<string, unknown>;
}

export interface MapRoute {
  name: string;
  title?: string;
  waypoints: MapLocation[];
  color?: string;
  width?: number;
  extra: Record<string, unknown>;
}

export interface MapView {
  center?: MapLocation;
  zoom?: number;
}

export interface MapDocument {
  kind: string;
  view?: MapView;
  markers: MapMarker[];
  areas: MapArea[];
  routes: MapRoute[];
  extra: Record<string, unknown>;
}

export class MapCodecError extends Error {
  constructor(message: string, public override readonly cause?: unknown) {
    super(message);
    this.name = 'MapCodecError';
  }
}

// ── MIME helpers ─────────────────────────────────────────────────────

function isJson(mime: string): boolean {
  return mime === 'application/json';
}
function isYaml(mime: string): boolean {
  return mime === 'application/yaml'
    || mime === 'application/x-yaml'
    || mime === 'text/yaml'
    || mime === 'text/x-yaml';
}

// ── Public API ───────────────────────────────────────────────────────

export function parseMap(body: string, mimeType: string): MapDocument {
  if (isJson(mimeType)) return parseMapJson(body);
  if (isYaml(mimeType)) return parseMapYaml(body);
  throw new MapCodecError(`Unsupported mime type for map: ${mimeType}`);
}

export function serializeMap(doc: MapDocument, mimeType: string): string {
  if (isJson(mimeType)) return serializeMapJson(doc);
  if (isYaml(mimeType)) return serializeMapYaml(doc);
  throw new MapCodecError(`Unsupported mime type for map: ${mimeType}`);
}

export function isMapMime(mimeType: string | null | undefined): boolean {
  if (!mimeType) return false;
  return isJson(mimeType) || isYaml(mimeType);
}

/** True when the location resolves directly without geocoding. */
export function hasCoords(loc: MapLocation | undefined): boolean {
  if (!loc) return false;
  return typeof loc.lat === 'number' && Number.isFinite(loc.lat)
    && typeof loc.lon === 'number' && Number.isFinite(loc.lon);
}

/** True when the location only carries a place name and needs to be
 *  geocoded before rendering. */
export function needsGeocode(loc: MapLocation | undefined): boolean {
  if (!loc) return false;
  if (hasCoords(loc)) return false;
  return typeof loc.place === 'string' && loc.place.trim().length > 0;
}

// ── JSON ─────────────────────────────────────────────────────────────

function parseMapJson(body: string): MapDocument {
  if (body.trim() === '') return emptyDoc();
  let parsed: unknown;
  try {
    parsed = JSON.parse(body);
  } catch (e) {
    throw new MapCodecError('Invalid JSON: ' + (e instanceof Error ? e.message : String(e)), e);
  }
  if (!isObject(parsed)) {
    throw new MapCodecError('Top-level JSON must be an object');
  }
  return promoteToMapDocument(unwrapJsonMeta(parsed));
}

function serializeMapJson(doc: MapDocument): string {
  const body = buildOnDiskBody(doc);
  return JSON.stringify(wrapJsonMeta(doc.kind || 'map', body), null, 2) + '\n';
}

// ── YAML ─────────────────────────────────────────────────────────────

function parseMapYaml(body: string): MapDocument {
  if (body.trim() === '') return emptyDoc();
  let merged: Record<string, unknown>;
  try {
    merged = parseYamlBody(body);
  } catch (e) {
    throw new MapCodecError('Invalid YAML: ' + (e instanceof Error ? e.message : String(e)), e);
  }
  return promoteToMapDocument(merged);
}

function serializeMapYaml(doc: MapDocument): string {
  return dumpYamlBody(doc.kind || 'map', buildOnDiskBody(doc));
}

// ── Promotion ────────────────────────────────────────────────────────

function emptyDoc(): MapDocument {
  return {
    kind: 'map',
    markers: [],
    areas: [],
    routes: [],
    extra: {},
  };
}

function promoteToMapDocument(obj: Record<string, unknown>): MapDocument {
  const kind = typeof obj.kind === 'string' ? obj.kind : '';
  const view = promoteView(obj.view);
  const markers = promoteMarkers(obj.markers);
  const areas = promoteAreas(obj.areas);
  const routes = promoteRoutes(obj.routes);

  const { kind: _k, view: _v, markers: _m, areas: _a, routes: _r, ...extra } = obj;
  const doc: MapDocument = {
    kind: kind || 'map',
    markers,
    areas,
    routes,
    extra,
  };
  if (view) doc.view = view;
  return doc;
}

function promoteView(raw: unknown): MapView | undefined {
  if (!isObject(raw)) return undefined;
  const center = promoteLocation(raw);
  const zoom = promoteInt(raw.zoom);
  if (!center && zoom === undefined) return undefined;
  const view: MapView = {};
  if (center) view.center = center;
  if (zoom !== undefined) view.zoom = zoom;
  return view;
}

// `label` is accepted as a `title` alias on every feature — that's
// the field LLMs naturally reach for. Likewise, missing `name` is
// auto-derived from the title (slugified) so a feature isn't lost
// when the model only writes a display label.
const MARKER_RESERVED = new Set(['name', 'title', 'label', 'place', 'lat', 'lon', 'color', 'description']);
const AREA_RESERVED = new Set(['name', 'title', 'label', 'points', 'color', 'fillOpacity']);
const ROUTE_RESERVED = new Set(['name', 'title', 'label', 'waypoints', 'color', 'width']);

function readTitle(r: Record<string, unknown>): string | undefined {
  return promoteString(r.title) ?? promoteString(r.label);
}

function autoName(kind: string, title: string | undefined, oneBasedIndex: number): string {
  if (title && title.trim().length > 0) {
    const slug = title.trim().toLowerCase()
      .replace(/[^a-z0-9]+/g, '-')
      .replace(/^-+|-+$/g, '');
    if (slug.length > 0) return slug;
  }
  return `${kind}_${oneBasedIndex}`;
}

function promoteMarkers(raw: unknown): MapMarker[] {
  if (!Array.isArray(raw)) return [];
  const out: MapMarker[] = [];
  let index = 0;
  for (const r of raw) {
    index++;
    if (!isObject(r)) continue;
    const loc = promoteLocation(r);
    if (!loc) continue;
    const title = readTitle(r);
    const name = promoteString(r.name) ?? autoName('marker', title, index);
    const marker: MapMarker = {
      name,
      location: loc,
      extra: {},
    };
    if (title) marker.title = title;
    const color = promoteString(r.color);
    if (color) marker.color = color;
    const description = promoteString(r.description);
    if (description) marker.description = description;
    for (const [k, v] of Object.entries(r)) {
      if (MARKER_RESERVED.has(k)) continue;
      marker.extra[k] = v;
    }
    out.push(marker);
  }
  return out;
}

function promoteAreas(raw: unknown): MapArea[] {
  if (!Array.isArray(raw)) return [];
  const out: MapArea[] = [];
  let index = 0;
  for (const r of raw) {
    index++;
    if (!isObject(r)) continue;
    const title = readTitle(r);
    const name = promoteString(r.name) ?? autoName('area', title, index);
    const area: MapArea = {
      name,
      points: promoteLocationList(r.points),
      extra: {},
    };
    if (title) area.title = title;
    const color = promoteString(r.color);
    if (color) area.color = color;
    const fillOpacity = promoteDouble(r.fillOpacity);
    if (fillOpacity !== undefined) area.fillOpacity = fillOpacity;
    for (const [k, v] of Object.entries(r)) {
      if (AREA_RESERVED.has(k)) continue;
      area.extra[k] = v;
    }
    out.push(area);
  }
  return out;
}

function promoteRoutes(raw: unknown): MapRoute[] {
  if (!Array.isArray(raw)) return [];
  const out: MapRoute[] = [];
  let index = 0;
  for (const r of raw) {
    index++;
    if (!isObject(r)) continue;
    const title = readTitle(r);
    const name = promoteString(r.name) ?? autoName('route', title, index);
    const route: MapRoute = {
      name,
      waypoints: promoteLocationList(r.waypoints),
      extra: {},
    };
    if (title) route.title = title;
    const color = promoteString(r.color);
    if (color) route.color = color;
    const width = promoteInt(r.width);
    if (width !== undefined) route.width = width;
    for (const [k, v] of Object.entries(r)) {
      if (ROUTE_RESERVED.has(k)) continue;
      route.extra[k] = v;
    }
    out.push(route);
  }
  return out;
}

function promoteLocationList(raw: unknown): MapLocation[] {
  if (!Array.isArray(raw)) return [];
  const out: MapLocation[] = [];
  for (const r of raw) {
    if (!isObject(r)) continue;
    const loc = promoteLocation(r);
    if (loc) out.push(loc);
  }
  return out;
}

function promoteLocation(map: Record<string, unknown>): MapLocation | undefined {
  const place = promoteString(map.place);
  const lat = promoteDouble(map.lat);
  const lon = promoteDouble(map.lon);
  if (!place && (lat === undefined || lon === undefined)) return undefined;
  const loc: MapLocation = {};
  if (place) loc.place = place;
  if (lat !== undefined) loc.lat = lat;
  if (lon !== undefined) loc.lon = lon;
  return loc;
}

function promoteString(raw: unknown): string | undefined {
  if (typeof raw !== 'string') return undefined;
  const trimmed = raw.trim();
  return trimmed.length === 0 ? undefined : trimmed;
}

function promoteDouble(raw: unknown): number | undefined {
  if (typeof raw !== 'number' || !Number.isFinite(raw)) return undefined;
  return raw;
}

function promoteInt(raw: unknown): number | undefined {
  if (typeof raw !== 'number' || !Number.isFinite(raw)) return undefined;
  return Math.round(raw);
}

// ── On-disk writer ───────────────────────────────────────────────────

function buildOnDiskBody(doc: MapDocument): Record<string, unknown> {
  const body: Record<string, unknown> = {};
  if (doc.view) {
    const view = viewToObject(doc.view);
    if (Object.keys(view).length > 0) body.view = view;
  }
  body.markers = doc.markers.map(markerToObject);
  body.areas = doc.areas.map(areaToObject);
  body.routes = doc.routes.map(routeToObject);
  for (const [k, v] of Object.entries(doc.extra)) {
    if (!(k in body)) body[k] = v;
  }
  return body;
}

function viewToObject(view: MapView): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  if (view.center) writeLocationInto(out, view.center);
  if (view.zoom !== undefined) out.zoom = view.zoom;
  return out;
}

function markerToObject(marker: MapMarker): Record<string, unknown> {
  const obj: Record<string, unknown> = { name: marker.name };
  if (marker.title !== undefined) obj.title = marker.title;
  writeLocationInto(obj, marker.location);
  if (marker.color !== undefined) obj.color = marker.color;
  if (marker.description !== undefined) obj.description = marker.description;
  for (const [k, v] of Object.entries(marker.extra)) {
    if (!(k in obj)) obj[k] = v;
  }
  return obj;
}

function areaToObject(area: MapArea): Record<string, unknown> {
  const obj: Record<string, unknown> = { name: area.name };
  if (area.title !== undefined) obj.title = area.title;
  obj.points = area.points.map(locationToObject);
  if (area.color !== undefined) obj.color = area.color;
  if (area.fillOpacity !== undefined) obj.fillOpacity = area.fillOpacity;
  for (const [k, v] of Object.entries(area.extra)) {
    if (!(k in obj)) obj[k] = v;
  }
  return obj;
}

function routeToObject(route: MapRoute): Record<string, unknown> {
  const obj: Record<string, unknown> = { name: route.name };
  if (route.title !== undefined) obj.title = route.title;
  obj.waypoints = route.waypoints.map(locationToObject);
  if (route.color !== undefined) obj.color = route.color;
  if (route.width !== undefined) obj.width = route.width;
  for (const [k, v] of Object.entries(route.extra)) {
    if (!(k in obj)) obj[k] = v;
  }
  return obj;
}

function locationToObject(loc: MapLocation): Record<string, unknown> {
  const obj: Record<string, unknown> = {};
  writeLocationInto(obj, loc);
  return obj;
}

function writeLocationInto(obj: Record<string, unknown>, loc: MapLocation): void {
  if (loc.place !== undefined) obj.place = loc.place;
  if (loc.lat !== undefined && loc.lon !== undefined) {
    obj.lat = loc.lat;
    obj.lon = loc.lon;
  }
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

File: server/vance-shared/src/main/java/de/mhus/vance/shared/document/kind/MapCodec.java
package de.mhus.vance.shared.document.kind;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.jspecify.annotations.Nullable;
import tools.jackson.core.JacksonException;
import tools.jackson.core.type.TypeReference;
import tools.jackson.databind.ObjectMapper;

/**
 * Parser and serialiser for {@code kind: map} document bodies —
 * markers, areas and routes on a Leaflet-rendered OpenStreetMap.
 * JSON and YAML only; markdown is intentionally not supported (same
 * rationale as {@code graph}).
 *
 * <p>Position is carried flat on markers ({@code place}/{@code lat}/
 * {@code lon} at the marker level, no nested {@code location:} object)
 * to match the way an LLM naturally emits such YAML. Area points and
 * route waypoints are lists of the same flat shape.
 *
 * <p>Spec: {@code specification/doc-kind-map.md}.
 */
public final class MapCodec {

    private static final ObjectMapper JSON = new ObjectMapper();
    private static final TypeReference<LinkedHashMap<String, Object>> JSON_MAP =
            new TypeReference<>() {};

    private MapCodec() {
        // utility class
    }

    public static MapDocument parse(String body, @Nullable String mimeType) {
        if (isJson(mimeType)) return parseJson(body);
        if (isYaml(mimeType)) return parseYaml(body);
        throw new KindCodecException("Unsupported mime type for map: " + mimeType);
    }

    public static String serialize(MapDocument doc, @Nullable String mimeType) {
        if (isJson(mimeType)) return serializeJson(doc);
        if (isYaml(mimeType)) return serializeYaml(doc);
        throw new KindCodecException("Unsupported mime type for map: " + mimeType);
    }

    public static boolean supports(@Nullable String mimeType) {
        return isJson(mimeType) || isYaml(mimeType);
    }

    // ── Mime ───────────────────────────────────────────────────────

    private static boolean isJson(@Nullable String mime) {
        return "application/json".equals(mime);
    }

    private static boolean isYaml(@Nullable String mime) {
        return "application/yaml".equals(mime)
                || "application/x-yaml".equals(mime)
                || "text/yaml".equals(mime)
                || "text/x-yaml".equals(mime);
    }

    // ── JSON / YAML ────────────────────────────────────────────────

    private static MapDocument parseJson(String body) {
        if (body.isBlank()) return MapDocument.empty();
        Map<String, Object> parsed;
        try {
            parsed = JSON.readValue(body, JSON_MAP);
        } catch (JacksonException e) {
            throw new KindCodecException("Invalid JSON: " + e.getOriginalMessage(), e);
        }
        if (parsed == null) throw new KindCodecException("Top-level JSON must be an object");
        return promoteToDocument(KindHeaderCodec.unwrapJsonMeta(parsed));
    }

    private static MapDocument parseYaml(String body) {
        if (body.isBlank()) return MapDocument.empty();
        return promoteToDocument(KindHeaderCodec.parseYamlBody(body));
    }

    private static String serializeJson(MapDocument doc) {
        Map<String, Object> wrapped = KindHeaderCodec.wrapJsonMeta(canonicalKind(doc), buildBody(doc));
        try {
            return JSON.writerWithDefaultPrettyPrinter().writeValueAsString(wrapped) + "\n";
        } catch (JacksonException e) {
            throw new KindCodecException("Failed to write JSON: " + e.getOriginalMessage(), e);
        }
    }

    private static String serializeYaml(MapDocument doc) {
        return KindHeaderCodec.dumpYamlBody(canonicalKind(doc), buildBody(doc));
    }

    // ── Promotion ──────────────────────────────────────────────────

    private static MapDocument promoteToDocument(Map<String, Object> obj) {
        Object kindRaw = obj.get("kind");
        String kind = (kindRaw instanceof String s) ? s : "";
        MapView view = promoteView(obj.get("view"));
        List<MapMarker> markers = promoteMarkers(obj.get("markers"));
        List<MapArea> areas = promoteAreas(obj.get("areas"));
        List<MapRoute> routes = promoteRoutes(obj.get("routes"));

        Map<String, Object> extra = new LinkedHashMap<>(obj);
        extra.remove("kind");
        extra.remove("view");
        extra.remove("markers");
        extra.remove("areas");
        extra.remove("routes");
        return new MapDocument(kind.isEmpty() ? "map" : kind, view, markers, areas, routes, extra);
    }

    private static @Nullable MapView promoteView(@Nullable Object raw) {
        if (!(raw instanceof Map<?, ?> map)) return null;
        MapLocation center = promoteLocation(map);
        Integer zoom = promoteInt(map.get("zoom"));
        if (center == null && zoom == null) return null;
        return new MapView(center, zoom);
    }

    private static List<MapMarker> promoteMarkers(@Nullable Object raw) {
        List<MapMarker> out = new ArrayList<>();
        if (!(raw instanceof List<?> list)) return out;
        int index = 0;
        for (Object r : list) {
            index++;
            if (!(r instanceof Map<?, ?> map)) continue;
            MapLocation loc = promoteLocation(map);
            if (loc == null) continue;
            // `label` is accepted as an alias for `title` — that's
            // what LLMs naturally reach for. Title wins on collision.
            String title = promoteString(map.get("title"));
            if (title == null) title = promoteString(map.get("label"));
            String name = promoteString(map.get("name"));
            if (name == null) name = autoName("marker", title, index);
            String color = promoteString(map.get("color"));
            String description = promoteString(map.get("description"));

            Map<String, Object> extra = new LinkedHashMap<>();
            for (Map.Entry<?, ?> e : map.entrySet()) {
                if (!(e.getKey() instanceof String key)) continue;
                if (isMarkerReservedKey(key)) continue;
                extra.put(key, e.getValue());
            }
            out.add(new MapMarker(name, title, loc, color, description, extra));
        }
        return out;
    }

    private static boolean isMarkerReservedKey(String key) {
        return switch (key) {
            case "name", "title", "label", "place", "lat", "lon", "color", "description" -> true;
            default -> false;
        };
    }

    private static List<MapArea> promoteAreas(@Nullable Object raw) {
        List<MapArea> out = new ArrayList<>();
        if (!(raw instanceof List<?> list)) return out;
        int index = 0;
        for (Object r : list) {
            index++;
            if (!(r instanceof Map<?, ?> map)) continue;
            List<MapLocation> points = promoteLocationList(map.get("points"));
            String title = promoteString(map.get("title"));
            if (title == null) title = promoteString(map.get("label"));
            String name = promoteString(map.get("name"));
            if (name == null) name = autoName("area", title, index);
            String color = promoteString(map.get("color"));
            Double fillOpacity = promoteDouble(map.get("fillOpacity"));

            Map<String, Object> extra = new LinkedHashMap<>();
            for (Map.Entry<?, ?> e : map.entrySet()) {
                if (!(e.getKey() instanceof String key)) continue;
                if (isAreaReservedKey(key)) continue;
                extra.put(key, e.getValue());
            }
            out.add(new MapArea(name, title, points, color, fillOpacity, extra));
        }
        return out;
    }

    private static boolean isAreaReservedKey(String key) {
        return switch (key) {
            case "name", "title", "label", "points", "color", "fillOpacity" -> true;
            default -> false;
        };
    }

    private static List<MapRoute> promoteRoutes(@Nullable Object raw) {
        List<MapRoute> out = new ArrayList<>();
        if (!(raw instanceof List<?> list)) return out;
        int index = 0;
        for (Object r : list) {
            index++;
            if (!(r instanceof Map<?, ?> map)) continue;
            List<MapLocation> waypoints = promoteLocationList(map.get("waypoints"));
            String title = promoteString(map.get("title"));
            if (title == null) title = promoteString(map.get("label"));
            String name = promoteString(map.get("name"));
            if (name == null) name = autoName("route", title, index);
            String color = promoteString(map.get("color"));
            Integer width = promoteInt(map.get("width"));

            Map<String, Object> extra = new LinkedHashMap<>();
            for (Map.Entry<?, ?> e : map.entrySet()) {
                if (!(e.getKey() instanceof String key)) continue;
                if (isRouteReservedKey(key)) continue;
                extra.put(key, e.getValue());
            }
            out.add(new MapRoute(name, title, waypoints, color, width, extra));
        }
        return out;
    }

    private static boolean isRouteReservedKey(String key) {
        return switch (key) {
            case "name", "title", "label", "waypoints", "color", "width" -> true;
            default -> false;
        };
    }

    /**
     * Derive a stable {@code name} when the on-disk feature has none.
     * LLMs reach for {@code label} (as display) and skip the
     * technical name; slugifying the title keeps the persisted
     * artifact human-readable, while {@code <kind>_<idx>} catches the
     * fully-untitled case.
     */
    private static String autoName(String kind, @Nullable String title, int oneBasedIndex) {
        if (title != null && !title.isBlank()) {
            String slug = title.trim().toLowerCase()
                    .replaceAll("[^a-z0-9]+", "-")
                    .replaceAll("^-+|-+$", "");
            if (!slug.isEmpty()) return slug;
        }
        return kind + "_" + oneBasedIndex;
    }

    private static List<MapLocation> promoteLocationList(@Nullable Object raw) {
        List<MapLocation> out = new ArrayList<>();
        if (!(raw instanceof List<?> list)) return out;
        for (Object r : list) {
            if (!(r instanceof Map<?, ?> map)) continue;
            MapLocation loc = promoteLocation(map);
            if (loc != null) out.add(loc);
        }
        return out;
    }

    /**
     * Pull {@code place}/{@code lat}/{@code lon} off the given map.
     * Returns {@code null} when neither a usable place name nor a
     * coordinate pair is present.
     */
    private static @Nullable MapLocation promoteLocation(Map<?, ?> map) {
        String place = promoteString(map.get("place"));
        Double lat = promoteDouble(map.get("lat"));
        Double lon = promoteDouble(map.get("lon"));
        if (place == null && (lat == null || lon == null)) return null;
        return new MapLocation(place, lat, lon);
    }

    private static @Nullable String promoteString(@Nullable Object raw) {
        if (raw instanceof String s) {
            String trimmed = s.trim();
            return trimmed.isEmpty() ? null : trimmed;
        }
        return null;
    }

    private static @Nullable Double promoteDouble(@Nullable Object raw) {
        if (raw instanceof Number n) {
            double d = n.doubleValue();
            return Double.isFinite(d) ? d : null;
        }
        return null;
    }

    private static @Nullable Integer promoteInt(@Nullable Object raw) {
        if (raw instanceof Number n) {
            double d = n.doubleValue();
            if (!Double.isFinite(d)) return null;
            return (int) Math.round(d);
        }
        return null;
    }

    // ── Body builder ───────────────────────────────────────────────

    private static Map<String, Object> buildBody(MapDocument doc) {
        Map<String, Object> body = new LinkedHashMap<>();
        if (doc.view() != null) {
            Map<String, Object> view = viewToMap(doc.view());
            if (!view.isEmpty()) body.put("view", view);
        }
        body.put("markers", markersToList(doc.markers()));
        body.put("areas", areasToList(doc.areas()));
        body.put("routes", routesToList(doc.routes()));
        for (Map.Entry<String, Object> e : doc.extra().entrySet()) {
            if (!body.containsKey(e.getKey())) body.put(e.getKey(), e.getValue());
        }
        return body;
    }

    private static Map<String, Object> viewToMap(MapView view) {
        Map<String, Object> out = new LinkedHashMap<>();
        if (view.center() != null) writeLocationInto(out, view.center());
        if (view.zoom() != null) out.put("zoom", view.zoom());
        return out;
    }

    private static List<Map<String, Object>> markersToList(List<MapMarker> markers) {
        List<Map<String, Object>> out = new ArrayList<>(markers.size());
        for (MapMarker m : markers) {
            Map<String, Object> obj = new LinkedHashMap<>();
            obj.put("name", m.name());
            if (m.title() != null) obj.put("title", m.title());
            writeLocationInto(obj, m.location());
            if (m.color() != null) obj.put("color", m.color());
            if (m.description() != null) obj.put("description", m.description());
            for (Map.Entry<String, Object> e : m.extra().entrySet()) {
                if (!obj.containsKey(e.getKey())) obj.put(e.getKey(), e.getValue());
            }
            out.add(obj);
        }
        return out;
    }

    private static List<Map<String, Object>> areasToList(List<MapArea> areas) {
        List<Map<String, Object>> out = new ArrayList<>(areas.size());
        for (MapArea a : areas) {
            Map<String, Object> obj = new LinkedHashMap<>();
            obj.put("name", a.name());
            if (a.title() != null) obj.put("title", a.title());
            obj.put("points", locationsToList(a.points()));
            if (a.color() != null) obj.put("color", a.color());
            if (a.fillOpacity() != null) obj.put("fillOpacity", a.fillOpacity());
            for (Map.Entry<String, Object> e : a.extra().entrySet()) {
                if (!obj.containsKey(e.getKey())) obj.put(e.getKey(), e.getValue());
            }
            out.add(obj);
        }
        return out;
    }

    private static List<Map<String, Object>> routesToList(List<MapRoute> routes) {
        List<Map<String, Object>> out = new ArrayList<>(routes.size());
        for (MapRoute r : routes) {
            Map<String, Object> obj = new LinkedHashMap<>();
            obj.put("name", r.name());
            if (r.title() != null) obj.put("title", r.title());
            obj.put("waypoints", locationsToList(r.waypoints()));
            if (r.color() != null) obj.put("color", r.color());
            if (r.width() != null) obj.put("width", r.width());
            for (Map.Entry<String, Object> e : r.extra().entrySet()) {
                if (!obj.containsKey(e.getKey())) obj.put(e.getKey(), e.getValue());
            }
            out.add(obj);
        }
        return out;
    }

    private static List<Map<String, Object>> locationsToList(List<MapLocation> locations) {
        List<Map<String, Object>> out = new ArrayList<>(locations.size());
        for (MapLocation loc : locations) {
            Map<String, Object> obj = new LinkedHashMap<>();
            writeLocationInto(obj, loc);
            out.add(obj);
        }
        return out;
    }

    /**
     * Write {@code place}/{@code lat}/{@code lon} into the given map.
     * Used for flat markers (location fields at the marker level) and
     * for entries inside {@code points}/{@code waypoints} arrays.
     */
    private static void writeLocationInto(Map<String, Object> obj, MapLocation loc) {
        if (loc.hasPlace()) obj.put("place", loc.place());
        if (loc.hasCoords()) {
            obj.put("lat", loc.lat());
            obj.put("lon", loc.lon());
        }
    }

    private static String canonicalKind(MapDocument doc) {
        return (doc.kind() == null || doc.kind().isBlank()) ? "map" : doc.kind();
    }
}
